### PR TITLE
feat: show extensions warning in css learning module index

### LINKED ISSUE #38353: Add a warning that browser extensions may cause issues
More than a few times, someone has had issues passing a challenge because of interference from a browser extension(s). Seems to mostly happen in the CSS section or other areas where users are changing styles. [Original discussion here](https://github.com/freeCodeCamp/freeCodeCamp/issues/37774). We could probably add a warning about this.

### PR DESCRIPTION
**Pull request by luqven**
# Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Closes #38353

## Changes
This PR creates an `<AdBlockWarning>` component and uses it in the `<Intro>` component. The warning is only visible if an ad-blocking extension is detected.

The ad-block warning consist of an absolutely positioned 'snack-bar' at the bottom of the page that can be dismissed for the duration of that user's session. Once dismissed, a `@@adBlockDismiss` key with the value of `'true'` is added to the `sessionStorage`.

The PR also adds a test to `/components/Intro/Intro.test.js` to ensure the warning is only visible when an ad-blocking extension is present.

## Screenshots:
Desktop
Mobile

### Precedent

The ad-blocking detection script mostly came from [David Walsh's](https://davidwalsh.name/detect-ad-blocker) blog post, but the unnecessary `Request` class instantiation has been removed.

### Up for discussion

If the preference is for this warning to be visible within _any_ page where an ad-blocking extension is detected, please advise so can include this component in the header instead and update the PR.

Obviously, some executive decisions were made regarding placement, font-size, and color. More than happy to change those things as well.

### CODE REVIEW

**Comment by RandellDawson** [comment]
@luqven Not sure why you are importing `node-fetch` here instead of just using `fetch`.

**Comment by luqven** [comment]
> @luqven Not sure why you are importing `node-fetch` here instead of just using `fetch`.

Brain fart 🤣 , I was getting a linting error since `.eslintrc.json` doesn't have `fetch` as a global variable and explicitly importing `node-fetch` makes the error go away. I will replace the import statement with a comment [signaling](https://eslint.org/docs/2.0.0/user-guide/configuring#specifying-globals) global `fetch` is being used in the file. Thanks for pointing this out!

ps: should there be a `"globals": { "fetch": false}` somewhere in the `.eslintrc.json` file to avoid this linting issue moving forward?

**Comment by ojeytonwilliams** [comment]
I think the warning should be shown for all users, not just those using ad block.  While we know there's a specific problem caused by uBlock, in principle any extension could stop challenges working.

Putting this warning in the intros will mean it gets lost if the intros get removed, which has been in discussion for a while: https://github.com/freeCodeCamp/freeCodeCamp/issues/37144.  I recommend either putting the warning in the first challenge of the css section or creating a cypress test. It should checks that the warning exists and have instructions about where it should go if the intros are removed.

**Comment by ojeytonwilliams** [comment]
> ps: should there be a "globals": { "fetch": false} somewhere in the .eslintrc.json file to avoid this linting issue moving forward?

It already is included as part of `"env": {"browser": true}` which is part of `@freecodecamp/eslint-config`.  When were you getting lint issues?

**Comment by raisedadead** [comment]
This is a great start.  

Quick feedback:

We do not show aggressive warnings, so we should probably put this somewhere in the challenge views for starters and also re-use our offline message style notification instead. 

@ahmadabdolsaheb please guide.

**Comment by moT01** [comment]
nice job @luqven! Although I think this is a little aggressive as well. And I'm not sure it's really what was discussed in the issue, perhaps there wasn't enough details given. But I like the initiative 😄 

I had in mind something a lot simpler, like this:
Screen Shot 2020-05-13 at 6 32 54 AM
Which is in [this file](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/client/src/pages/learn/responsive-web-design/basic-css/index.md) - but perhaps that's too simple and hidden.

I don't mind the pop up warning if others a okay with it. I would move it to the page I linked, show it for everybody and not just those using ad-blockers like @ojeytonwilliams mentioned, and "use our offline message style notification instead" like @raisedadead mentioned. If you do those three, I think it might be good to go. But you may want to wait for @ahmadabdolsaheb to chime in.

**Comment by raisedadead** [comment]
Should we also change the verbiage to say something along the lines "if tests fail,  try disabling your extensions first." ?

**Comment by luqven** [comment]
Hey all, thanks for the great feedback! 

Am I correct in assuming the "offline message style notification" refers to `/components/Flash/flash.css`? If so, could be something as simple as adding the following to `/pages/learn/responsive-web-design/basic-css/index.md`:

```html
<div role="alert" class="fade alert alert-info show flash-message-enter-active">
    <p>
        Note: if tests fail, try disabling your extensions first.
    </p>
</div>
```
Bootstrap warning screenshot

If that "offline-warning" refers to `/components/OfflineWarning/OfflineWarning.js` instead, the message would could like:

```html
<div class="offline-warning" style="margin-bottom: 15px">
        Note: if tests fail, try disabling your extensions first.
</div>
```

Offline warning screenshot

Please advise as to which style is preferred, a) the bootstrap reliant`flash` or b) the `offline-warning`.

**Comment by raisedadead** [comment]
I think the bootstrap version is good enough.

Also, I was not clear on the exact message (was making more of a suggestion) so think we would want something like:

> Note: Some browser extensions may interfere with elements on the page. If the tests fail, try disabling your extensions for the most reliable experience.

**Comment by luqven** [comment]
> note: force-pushed to remove commits that created now irrelevant `<AdBlockWarning>` React components.

The warning message is now

-  located in the page linked by @moT01 
-  shown for everybody and not just those using ad-blockers like @ojeytonwilliams mentioned 
- using the repo's notification styles like @raisedadead mentioned.
- displaying the text suggest by @raisedadead 

Thank again to all for the feedback and advice 😄

File: cypress/integration/learn/responsive-web-design/basic-css/index.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  'Create a custom CSS Variable',
+  'Use a custom CSS Variable',
+  'Attach a Fallback value to a CSS Variable',
+  'Improve Compatibility with Browser Fallbacks',
+  'Inherit CSS Variables',
+  'Change a variable for a specific area',
+  'Use a media query to change a variable'
+];
+
+const warningMessage = 'Note: Some browser extensions may';
```
Comment (ojeytonwilliams) [inline cypress/integration/learn/responsive-web-design/basic-css/index.js:59]:
```suggestion
const warningMessage =
  'Note: Some browser extensions may interfere with elements on the page. ' +
  'If the tests fail, try disabling your extensions for the most reliable ' +
  'experience.';
```
It's minor, but we might as well check for the entire message.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
I've made a suggestion about the warning message, but other than that, this looks really good.

File: cypress/integration/learn/responsive-web-design/basic-css/index.js (hunk 1 above)
Comment (luqven) [reply cypress/integration/learn/responsive-web-design/basic-css/index.js:59]: Gotcha, will update to check for entire message 👍🏻

Comment (luqven) [reply cypress/integration/learn/responsive-web-design/basic-css/index.js]: @ojeytonwilliams Let me know if anything else needs addressing, above should be fixed in latest push 👍🏻

**Review by ojeytonwilliams — APPROVED** [review]
Everything looks great.  Nice work!

**Comment by ojeytonwilliams** [comment]
@moT01 can you give this one last look?

**Review by moT01 — COMMENTED** [review]
Overall, this LGTM 🎉 - Do you guys think it's a good idea to hard code those lesson names in the tests? If a new lesson gets added, it won't be in there. If we want it in there, it's just one more thing to try and remember to add. I really don't see a problem with it right now since we won't be adding any new lessons there, just wondering what you all thought.

**Comment by ojeytonwilliams** [comment]
My thinking was that the basic css section is not supposed to change any more, so this should not be a problem.  It's effectively an e2e snapshot.

**Review by raisedadead — APPROVED** [review]

**Comment by raisedadead** [comment]
#### :sparkles: :sparkles: **AWESOME! LGTM.** :sparkles: :sparkles: 

Thanks @luqven, for this pull request and contributing to the code-base for the first time.
We will be looking forward for more contributions in future.

Cheers and happy contributing!

**Pull request merged**